omniledger/service/service.go

```diff
@@ -217,35 +218,33 @@ func (s *Service) SetKeyValue(req *SetKeyValue) (*SetKeyValueResponse, error) {
 	}, nil
 }
 
-// GetValue looks up the key in the given skipchain and returns the
-// corresponding value.
-func (s *Service) GetValue(req *GetValue) (*GetValueResponse, error) {
+// GetProof searches for a key and returns a proof of the
+// presence or the absence of this key.
+func (s *Service) GetProof(req *GetProof) (resp *GetProofResponse, err error) {
 	if req.Version != CurrentVersion {
 		return nil, errors.New("version mismatch")
 	}
-
-	key := append(append(req.Kind, []byte(":")...), req.Key...)
-	value, sig, err := s.getCollection(req.SkipchainID).GetValue(key)
+	latest, err := s.db().GetLatest(s.db().GetByID(req.ID))
 	if err != nil {
-		return nil, errors.New("couldn't get value for key: " + err.Error())
+		return
 	}
-	return &GetValueResponse{
-		Version:   CurrentVersion,
-		Value:     &value,
-		Signature: &sig,
-	}, nil
-}
-
-func getKey(tx *Transaction) []byte {
-	return append(append(tx.Kind, []byte(":")...), tx.Key...)
+	proof, err := NewProof(s.getCollection(req.ID), s.db(), latest.Hash, req.Key)
+	if err != nil {
+		return
+	}
+	resp = &GetProofResponse{
+		Version: CurrentVersion,
+		Proof:   *proof,
```

Comment:
curious about how proof is serialised. It is the proof from Service, which contains a `collection.Proof`. But `collection.Proof` only contain private fields. Wouldn't verification fail if the proof is sent over the network?

Reply:
Good catch! I added an `api_test.go` and created a PR to make this one lighter: #26